clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -556,8 +571,10 @@ class DeviceFunctionTracker {
   ~DeviceFunctionTracker() {
     DiagDeviceFunction Diagnoser{SemaRef, RecursiveFunctions};
     for (const FunctionDecl *FD : DeviceFunctions)
-      if (const FunctionDecl *Def = FD->getDefinition())
+      if (const FunctionDecl *Def = FD->getDefinition()) {
+        Diagnoser.SetIsConstexprFD(FD);
```

Comment:
So I don't think this is right, it still breaks in cases where the function is constexpr, but it is called in a non-constexpr context.

Consider:

````
constexpr void foo(int I) { if (I) foo(I -1); }
constexpr void bar() { foo(5); } // Should also diagnose.

...
kernel_single_task<class F>([]() { 
foo(5);  // Should still diagnose.
bar();
});
````

Checking the constexprness of a function isn't going to do the right thing, what you care about is the calling context, not the function itself.